cache/test_purge.py

```diff
@@ -35,8 +35,124 @@ def chains(self):
             chains.proxy(method='HEAD', uri=uri),
             chains.cache(method='GET', uri=uri),
             chains.cache(method='HEAD', uri=uri)
-            ]
+        ]
         return result
 
     def test_purge(self):
         self.generic_test_routine(self.config, self.chains())
+
+    def purge_get(self, uri, method='GET'):
+        c = chains.proxy(uri=uri)
+        c.request.method = 'PURGE'
+        c.request.headers['X-Tempesta-Cache'] = method
+        c.request.update()
+        # A response body must be empty
+        c.response.headers['Content-Length'] = '0'
+        c.response.body = ''
+        c.response.update()
+        # Our X-header goes to an upstream
+        c.fwd_request.headers['X-Tempesta-Cache'] = method
+        return c
+
+    def test_purge_get_basic(self):
+        # Basic PURGE use.
+        uri = '/page.html'
+        ch = [
+            # All cacheable method to the resource must be cached
+            chains.proxy(method='GET', uri=uri),
+            chains.proxy(method='HEAD', uri=uri),
+            chains.cache(method='GET', uri=uri),
+            chains.cache(method='HEAD', uri=uri),
+            # PURGE + GET works like a cache update, so all following requests
+            # must be served from the cache.
+            self.purge_get(uri),
+            chains.cache(method='GET', uri=uri),
+            # Note that due to how Tempesta handles HEAD this doesn't help us
+            # with HEAD pre-caching.
+            chains.proxy(method='HEAD', uri=uri),
+            chains.cache(method='HEAD', uri=uri),
+        ]
+        self.generic_test_routine(self.config, ch)
+
+    def test_purge_get_update(self):
+        # Return a new response from the upstream to the PURGE+GET request
+        # pair, make sure we receive this new response from subsequent
+        # requests.
+        uri = '/page.html'
+        page = 'New page text!\n'
+        ch = [
+            chains.proxy(method='GET', uri=uri),
+            chains.cache(method='GET', uri=uri),
+            self.purge_get(uri),
+            chains.cache(method='GET', uri=uri),
+        ]
+        for c in (ch[2].server_response, ch[3].response):
+            c.body = page
+            c.headers['Content-Length'] = len(page)
+            c.update()
+        self.generic_test_routine(self.config, ch)
+
+    def test_useless_x_tempesta_cache(self):
+        # Send "X-Tempesta-Cache" with a GET request, and make sure it doesn't
+        # affect anything.
```

Comment:
I didn't the comment: why doesn't PURGE affect anything - is it due to uncacheable response or non-existing cache entry? Could you please elaborate on the comment to make it clear what it does and what it checks

Reply:
It doesn't do anything here, because the request isn't a PURGE to begin with. So FW handles this case as an ordinary GET with a "useless" header. This is the first bullet point from #1516.